skrifa/src/meta/charmap.rs

```diff
@@ -0,0 +1,420 @@
+        if let Ok(cmap) = font.cmap() {
+            MappingSelection::new(&cmap).mapping_index
+        } else {
+            Default::default()
+        }
+    }
+
+    /// Creates a new character map for the given font using the tables referenced by
+    /// the precomputed indices.
+    ///
+    /// The font should be the same as the one used to construct this object.
+    pub fn charmap<'a>(&self, font: &impl TableProvider<'a>) -> Charmap<'a> {
+        if let Ok(cmap) = font.cmap() {
+            let records = cmap.encoding_records();
+            let data = cmap.offset_data();
+            Charmap {
+                codepoint_subtable: self
+                    .codepoint_subtable
+                    .and_then(|index| get_subtable(data, records, index))
+                    .and_then(SupportedSubtable::new)
+                    .map(|subtable| CodepointSubtable {
+                        subtable,
+                        is_symbol: self.codepoint_subtable_is_symbol,
+                    }),
+                variant_subtable: self
+                    .variant_subtable
+                    .and_then(|index| get_subtable(data, records, index))
+                    .and_then(|subtable| match subtable {
+                        CmapSubtable::Format14(cmap14) => Some(cmap14),
+                        _ => None,
+                    }),
+            }
+        } else {
+            Default::default()
+        }
+    }
+}
+
+fn get_subtable<'a>(
+    data: FontData<'a>,
+    records: &[EncodingRecord],
+    index: u16,
+) -> Option<CmapSubtable<'a>> {
+    records
+        .get(index as usize)
+        .and_then(|record| record.subtable(data).ok())
+}
+
+#[derive(Clone)]
+struct CodepointSubtable<'a> {
+    subtable: SupportedSubtable<'a>,
+    is_symbol: bool,
+}
+
+impl<'a> CodepointSubtable<'a> {
+    fn map(&self, codepoint: u32) -> Option<GlyphId> {
+        self.map_impl(codepoint).or_else(|| {
+            if self.is_symbol {
+                self.map_impl(Self::shift_symbol_to_pua(codepoint)?)
+            } else {
+                None
+            }
+        })
+    }
+
+    fn map_impl(&self, codepoint: u32) -> Option<GlyphId> {
+        match &self.subtable {
+            SupportedSubtable::Format4(subtable) => subtable.map_codepoint(codepoint),
+            SupportedSubtable::Format12(subtable) => subtable.map_codepoint(codepoint),
+        }
+    }
+
+    fn shift_symbol_to_pua(codepoint: u32) -> Option<u32> {
+        // From HarfBuzz:
+        // For symbol-encoded OpenType fonts, we duplicate the
+        // U+F000..F0FF range at U+0000..U+00FF.  That's what
+        // Windows seems to do, and that's hinted about at:
+        // https://docs.microsoft.com/en-us/typography/opentype/spec/recom
+        // under "Non-Standard (Symbol) Fonts".
+        // See <https://github.com/harfbuzz/harfbuzz/blob/453ded05392af38bba9f89587edce465e86ffa6b/src/hb-ot-cmap-table.hh#L1595>
+        if codepoint <= 0x00FF {
```

Comment:
Suggestion: take &self here and then make the if statement self.is_symbol && codepoint...

That lets you change map to something like self.map_impl(codepoint).or_else(Self::shift_symbol_to_pua)

Reply:
We can't quite get the clean syntax you suggest since `shift_symbol_to_pua` will require two arguments. Inlining the code would simplify things though.